### ISSUE TITLE
Multiple Entry Points each with Multiple Formats (with JS API)

### ISSUE THREAD

**Issue by agilgur5** [rollup/rollup#3422]
Documentation Is:

- [ ] Missing
- [ ] Needed
- [x] Confusing
- [ ] Not Sure?

### Please Explain in Detail...

I'm not entirely sure if what I'm asking about is possible, this might be a feature request. Per the title, I have a few entry points, and each of those needs to output multiple formats, and chunks should be shared between them.

So something like:
```javascript
rollup({
  input: ['src/ponyfill.js', 'src/polyfill.js'],
  output: [{
    dir: 'dist/esm/'
    format: 'esm'
  }, {
    dir: 'dist/cjs/'
    format: 'cjs'
  }]
}) 
```

Where `src/polyfill.js` imports `src/ponyfill.js`.

And, ideally would like to be able to have per-output plugins too, e.g. for minified vs. non-minified bundle.

This is related to #3325, as:
1. Can't do multiple configs via the API
2. `bundle.write` only takes one `OutputOptions`, not an array of `OutputOptions`, so I think this isn't possible via the API

Based on 2. I think I'd currently have to run `rollup` and `bundle.write` once on _each_ format right now, which is pretty time-consuming as none of the work is re-used between formats.

This is something I'm running into in with supporting multiple entries in TSDX: https://github.com/jaredpalmer/tsdx/pull/367 . Currently `rollup` is ran on each format _and_ on each input, which is very time-consuming and, more problematically, no code is shared between the two bundles.

Also seems related to the quite older #1150

### Your Proposal for Changes

Clarify if this is possible (with the API and CLI). If not, what are the current workarounds? Running `rollup` for each format with multiple inputs?

**Comment by lukastaegert** [comment]
> Based on 2. I think I'd currently have to run rollup and bundle.write once on each format right now

Please look at how the JS API works: https://rollupjs.org/guide/en/#rolluprollup
Running `rollup` generates a bundle object, but you can run `bundle.generate` or `bundle.write` as often as you want from the same bundle object. This is all the reuse that you are going to get: The heavy analysis including tree-shaking has been done in `rollup.rollup`; `bundle.generate/write` only does format specific things and generates the output.

This is how JS API users should generate multiple outputs. Rollup CLI does just that: https://github.com/rollup/rollup/blob/f458cbf6cb8cfcc1678593d8dc595e4b8757eb6d/cli/run/build.ts#L66

As this is apparently confusing to many, how could we improve the section on the JS API? Ideally, a PR to improve this would help everyone. If you think there is more you do not understand to feel comfortable doing this, please ask. A "newcomer" has a much better chance of writing something that will help other newcomers.

So going back to your example if you really do not want to or cannot use a config file, you would do

```js
const bundle = await rollup({input: ['src/ponyfill.js', 'src/polyfill.js']});
await Promise.all([
  bundle.write({
    dir: 'dist/esm'
    format: 'esm'
  }),
  bundle.write({
    dir: 'dist/cjs'
    format: 'cjs'
  })
])
```

> And, ideally would like to be able to have per-output plugins too

https://rollupjs.org/guide/en/#using-output-plugins
Did you read this?

**Comment by lukastaegert** [comment]
Also if you find the documentation on code-splitting lacking, maybe this can shed a little more light: https://levelup.gitconnected.com/code-splitting-for-libraries-bundling-for-npm-with-rollup-1-0-2522c7437697

If you have suggestions how the documentation on rollupjs.org could be improved with regard to how code-splitting is explained, please go ahead.

**Comment by agilgur5** [comment]
Thanks a bunch for the quick and detailed response!!

> Please look at how the JS API works: https://rollupjs.org/guide/en/#rolluprollup
> Running `rollup` generates a bundle object, but you can run `bundle.generate` or `bundle.write` as often as you want from the same bundle object. This is all the reuse that you are going to get: The heavy analysis including tree-shaking has been done in `rollup.rollup`; `bundle.generate/write` only does format specific things and generates the output.

Welp, I've read the docs like 20+ times but didn't realize I could reuse the same bundle object 😕 😮 

> As this is apparently confusing to many, how could we improve the section on the JS API? Ideally, a PR to improve this would help everyone. If you think there is more you do not understand to feel comfortable doing this, please ask. A "newcomer" has a much better chance of writing something that will help other newcomers.

For sure, I'll have to think about this. I think the main reason it's confusing is that the JS API and the CLI don't have identical interfaces. The docs focus around the CLI and the API's docs feel comparatively sparse.

The other thing that would be useful is more examples around these advanced (but not necessarily uncommon) use cases. If I saw once the bundle being reused, that would've been a big hint. Most examples use the CLI. Perhaps a section that shows the identical CLI configs done with the API would be optimal.

> or cannot use a config file

TSDX is a CLI tool that wraps `rollup` 😅 

> https://rollupjs.org/guide/en/#using-output-plugins
> Did you read this?

Yep, I even read your original PR. I meant being able to use that feature in addition to multiple entries and formats with the API, sorry if that was miscommunicated. Given the bundle re-usage I think I get how I'd use them now.

> Also if you find the documentation on code-splitting lacking, maybe this can shed a little more light: https://levelup.gitconnected.com/code-splitting-for-libraries-bundling-for-npm-with-rollup-1-0-2522c7437697

I think I read this 3 or 4 months ago, but maybe I didn't read it all the way to the end since some of it feels new (the advanced stuff is only at the end, so I might've not reached it before clicking away). Also didn't realize you were the author of that tutorial.

<hr>

Think this can be closed out unless you want to leave it open until more docs are added

**Comment by lukastaegert** [comment]
Well, looking at the docs I agree that a lot remains a mystery. As I do not want this to become a support forum but rather hope things can be improved once for everyone, I would leave this open until the section has been improved.

**Comment by agilgur5** [comment]
Ok, so I think I've figured mostly everything out now. The only remaining question I have is around UMD names -- you're limited to one name, despite being able to have an array of inputs. Does multi-entry not currently work with UMD?

**Comment by lukastaegert** [comment]
Yes, UMD does not support code-splitting because there is no mechanism to load dependencies in UMD. Thus one UMD output is always one bundle, and you can provide one name per output. There are some ideas to make some basic code-splitting without dynamic imports possible for UMD or IIFE but they are not actively pursued.

**Comment by frank-dspeed** [comment]
We should add compat to the rollup function so that when it gets called with a array or output options array.

to call multiple times rollupInternal this way api and cli would be the same behavior

```js
function rollup(rawRollupInputConfig) {
    const config = await rawRollupInputConfig
    return Array.isArray(config) ? Promise.all(config.map(rollupInternal)) : rollupInternal(config)
}
// And maybe then also modify bundle generate and bundle write to return arrays
```

**Comment by frank-dspeed** [comment]
i use this helpers internal 

```js
import rollup from 'rollup'
async buildFromConfig(rawRollupInputConfig) {
    const config = await rawRollupInputConfig
    return Array.isArray(config) ? Promise.all(config.map(build)) : build(config)
}


async function build(inputOptions, warnings, silent = false) {
    const bundle = await rollup(inputOptions);
    return Array.isArray(inputOptions.output) ? Promise.all(inputOptions.output.map(bundle.write)) : bundle.write(inputOptions.output)
}
```
